src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -239,7 +239,19 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                         {
                             frameReceived = true;
                             Log.Http2FrameReceived(ConnectionId, _incomingFrame);
-                            await ProcessFrameAsync(application, framePayload);
+
+                            try
+                            {
+                                await ProcessFrameAsync(application, framePayload);
+                            }
+                            catch (Http2StreamErrorException ex)
+                            {
+                                Log.Http2StreamError(ConnectionId, ex);
+                                // The client doesn't know this error is coming, allow draining additional frames for now.
+                                AbortStream(_incomingFrame.StreamId, new IOException(ex.Message, ex));
+
+                                await _frameWriter.WriteRstStreamAsync(ex.StreamId, ex.ErrorCode);
+                            }
```

Comment:
```suggestion

                                // Immediately resume reading frames after aborting HTTP/2 stream. This is important
                                // because other frames from other streams could be in the current buffer
                                // and we don't want to delay reading them until the next incoming read/heartbeat
                            }
```

Reply:
It's not just frames for other streams, even frames for the current stream need to continue to be processed. Yes we aborted the stream, but we must continue to decode headers, drain data, update connection windows, etc.

"This is important because additional frames could be in the current buffer."